### PR TITLE
Error definitions

### PR DESCRIPTION
**Pull request by bgoldstone**
Part of #4681.

# Pydantic Error Definitions V2

- [x] Added PydanticErrorMixin Class
- [X] Added constructor methods for each error type

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Looks good, but better to use a common mixin for all the error types.

**Comment by samuelcolvin** [comment]
Looks like a good start.

Please can you remove the default PR body and put some details on that you're working on.

**Comment by bgoldstone** [comment]
This should be good for Issue #4681

**Review by samuelcolvin — COMMENTED** [review]
This is a good start, now we need to set the errors everywhere that these errors are used.

**Comment by samuelcolvin** [comment]
please update.

File: pydantic/errors.py [hunk 1]
```diff
@@ -3,17 +3,47 @@
...
     """
 
+    def __str__(self) -> str:
+        return f'{self.__class__.__name__}(type="{self.type}", message="{self.message}")'
 
-class PydanticUndefinedAnnotation(NameError):
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def init_variables(self, type: str, message: str) -> None:
```
[the remark this answers is not in the document]
Comment (bgoldstone) [reply pydantic/errors.py:17]: is the **\*** supposed to be there for kwargs?

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/errors.py:17]: Yes, if `message` is included, it should be explicitly named.

[the remark this answers is not in the document]
Comment (bgoldstone) [reply pydantic/errors.py]: Do we have a list of errors @samuelcolvin? I saw some errors in an enum [here](https://github.com/pydantic/pydantic-core/blob/main/src/errors/types.rs)

[the remark this answers is not in the document]
Comment (bgoldstone) [reply pydantic/errors.py]: also would the ErrorCodes be a List[]?

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/errors.py]: these errors are completely different from the validation errors. The point is to add a code to each one (doesn't really matter what it is), so we can add a URL to the error and include more information on what might have gone wrong.

E.g. [here](https://github.com/pydantic/pydantic/blob/2e17c85810fd7e335422dcbf2525d019cfeb5d8c/pydantic/decorator.py#L66) you might use the code `arg-names`.

But don't worry too much, just use some short names and we can always change them before release.

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/errors.py]: signature of `__init__` would be

```py
def __init__(self, code: ErrorCodes, *, message: Optional[str] = None):
```

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/errors.py]: `ErrorCodes` is a `Literal`

**Comment by bgoldstone** [comment]
I think that is what you are looking for. Please let me know what you think. Thanks for the help! @samuelcolvin

**Comment by bgoldstone** [comment]
Is there anything else you would like modified @samuelcolvin?

**Comment by samuelcolvin** [comment]
We need tests passing, and as I said above, we need to actually use these everywhere.

**Comment by bgoldstone** [comment]
When running make I keep getting these errors
```make
pydantic\errors.py:25: error: Argument 1 for "super" must be a type object; got a non-type instance  [arg-type]
pydantic\errors.py:27: error: Argument 1 for "super" must be a type object; got a non-type instance  [arg-type]
```

I am not sure how else to initialize the TypeError/NameError constructors. Does anyone have any ideas?

**Comment by samuelcolvin** [comment]
sorry for the slow reply, did you fix your issue?

File: pydantic/errors.py [hunk 2]
```diff
@@ -1,19 +1,46 @@
...
+    Pydantic Error Mixin for common functions
     """
 
+    def __str__(self) -> str:
+        return f'{self.__class__.__name__}(code={self.code!r}, message={self.message!r})'
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def __init__(self, code: str, *, message: Optional[str] = None) -> None:
```
Comment (samuelcolvin) [inline pydantic/errors.py:19]: init should come first.

Comment (bgoldstone) [reply pydantic/errors.py]: Seems like tests keeps failing on tests/test_forward_ref.py

Comment (bgoldstone) [reply pydantic/errors.py]: Tried adding PydanticErrors to __init__ but it keeps failing as the linter says 
```make
 pydantic\__init__.py:7:21: F401 `.fields.Field` imported but unused
Found 1 error(s).
```

but if I remove Field it says 
```make
E   ImportError: cannot import name 'Field' from 'pydantic' (C:\Users\bgoldstone\Development\pydantic\pydantic\__init__.py)
```

Not sure if you have any suggestions... @samuelcolvin

**Comment by samuelcolvin** [comment]
I see some emails with questions, but I can't see them here.

If you need help, please ask here.

**Comment by bgoldstone** [comment]
> I see some emails with questions, but I can't see them here.
> 
> If you need help, please ask here.

Most of the errors seem to be occurring here @samuelcolvin
```make
                                                                      Summary of Failures                                                                         
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓   
┃  File                       ┃  Function                                                                 ┃  Function Line  ┃  Error Line  ┃  Error           ┃   
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩   
│  tests\test_forward_ref.py  │  test_forward_ref_auto_update_no_model                                    │  40             │  67          │  AssertionError  │   
│  tests\test_forward_ref.py  │  test_undefined_types_warning_1a_raised_by_default_2a_future_annotations  │  891            │  892         │  AssertionError  │   
│  tests\test_forward_ref.py  │  test_undefined_types_warning_1a_raised_by_default_2b_forward_ref         │  905            │  906         │  AssertionError  │   
└─────────────────────────────┴───────────────────────────────────────────────────────────────────────────┴─────────────────┴──────────────┴──────────────────┘
```
**test/test_forward_ref.py 67:69**
```py
 # but Foo is not ready to use
    with pytest.raises(PydanticUserError, match='`Foo` is not fully defined, you should define `Bar`,'):
        module.Foo(a={'b': {'a': {}}})
```
Does this need to be changed from match to "code"?

**Comment by bgoldstone** [comment]
> The new error messages look weird. Let me finish the thing I'm currently working on, and I'll look into it.

Sounds good @samuelcolvin

**Comment by samuelcolvin** [comment]
There errors here are because `str(name_error)` returns the name which we rely on when formatting `warning_string`.

you could probably add

```py
@property
def name(self) -> str:
    self.message
```

to `PydanticUndefinedAnnotation`. Then Like I said before, you need to change the calls to these error messages to set code correctly.

**Comment by bgoldstone** [comment]
> There errors here are because `str(name_error)` returns the name which we rely on when formatting `warning_string`.
> 
> you could probably add
> 
> ```python
> @property
> def name(self) -> str:
>     self.message
> ```
> 
> to `PydanticUndefinedAnnotation`. Then Like I said before, you need to change the calls to these error messages to set code correctly.

Yes. I will work on that. I am still not too familiar with the code base, but I am learning more as I go @samuelcolvin.

**Comment by samuelcolvin** [comment]
Thanks so much for this.

**Comment by bgoldstone** [comment]
> Thanks so much for this.

No problem. Thanks for helping me with my first PR! Hopefully, I can be able to help with more PRs.

**Pull request merged**